src/uct/tcp/tcp_cm.c

```diff
@@ -558,15 +583,55 @@ ucs_status_t uct_tcp_cm_conn_start(uct_tcp_ep_t *ep)
         }
     }
 
-    status = uct_tcp_cm_send_event(ep, UCT_TCP_CM_CONN_REQ);
+    return uct_tcp_cm_conn_complete(ep, NULL);;
+}
+
+static ucs_status_t uct_tcp_cm_magic_number_handler_io_err_cb(void *arg,
+                                                              int io_errno)
+{
+    ucs_debug("tcp_ep %p: failed to receive magic number: %s",
+              arg, strerror(io_errno));
+    return UCS_OK;
+}
+
+/* This function is called from async thread */
+static void uct_tcp_cm_magic_number_handler(int fd, void *arg)
+{
+    uct_tcp_ep_t *ep       = arg;
+    uct_tcp_iface_t *iface = ucs_derived_of(ep->super.super.iface,
+                                            uct_tcp_iface_t);
+    ucs_status_t status;
+    uint64_t magic_number;
+
+    UCS_ASYNC_BLOCK(iface->super.worker->async);
+    status = ucs_async_remove_handler(fd, 1);
+    UCS_ASYNC_UNBLOCK(iface->super.worker->async);
     if (status != UCS_OK) {
-        return status;
+        ucs_warn("failed to remove the magic number handler for "
+                 "client socket fd=%d", fd);
     }
 
-    uct_tcp_cm_change_conn_state(ep, UCT_TCP_EP_CONN_STATE_WAITING_ACK);
+    status = ucs_socket_recv(fd, &magic_number, sizeof(magic_number),
```

Comment:
is it possible we will receive only partial number and get UCS_ERR_NO_PROGRESS here?